### PR TITLE
Initial support for pytest

### PR DESCRIPTION
**Pull request by theHamsta**
Class support is missing, since the method selection is right now far to naive (just takes any class and any method, nested or not). Would just be `({path, class_name, closest_function}, '::')`.

-s is an opinionated setting that makes sense when you debug a single method (it deactivates the suppression of stdout).

I would recommend to have this not as a python specific plugin and just create the typical `query/lang/test.scm` for each language. Within a language one could have queries per test runner and also test whether a function for pytest starts with the magic `test_*`.

You can just have nvim-treesitter a dependency or copy the required function from there (`is_in_node_range`, `select_best_match`). nvim-treesitter-textobjects selects the smallest textobject than contains the current cursor positions. A similar strategy could be applied to get the hierarchy of the functions foo_class::test_foo vs test_foo.

### CODE REVIEW

**Comment by mfussenegger** [comment]
Thank you for the contribution, that was quick :)


> I would recommend to have this not as a python specific plugin and just create the typical query/lang/test.scm for each language.


It crossed my mind to create some kind of nvim-dap-polyglot plugin instead.
I went back and forth trying to decide and in the end decided to make it language specific.

Couple of reasons:

- Doesn't put the burden on me to look for other language maintainers or review PRs that are specific to languages I don't use
- Leaves the room open for more language specific extensions like I already created with nvim-jdtls. Extensions that are way bigger in scope.
- Doesn't constrain the plugin to provide only functionality that is a common subset of functionality that can be provided between all debug adapters. For example debugpy also provides the possibility to debug a code snippet - I started exposing that [here](https://github.com/mfussenegger/nvim-dap-python/pull/2)
- I doubt that will be a lot of complex code that could be shared across languages. Sure the `test_method` implementation is likely to look similar across languages, but a bit of code duplication is not *that* bad.

File: lua/dap-python.lua [hunk 1]
```diff
@@ -2,6 +2,9 @@ local api = vim.api
 local M = {}
 
 
+M.test_runner = 'unittest'
```
Comment (mfussenegger) [inline lua/dap-python.lua:5]: Do you think it should default to `pytest` if there is a `pytest` executable available? Afaik the runner should be able to run `unittest` style tests too?

**Review by mfussenegger — APPROVED** [review]

File: lua/dap-python.lua (hunk 1 above)
Comment (theHamsta) [reply lua/dap-python.lua:5]: ok, `vim.fn.executable('pytest') == 1 and 'pytest' or 'unittest'`

**Comment by theHamsta** [comment]
But the query files could still be in the standard directory to be overwritable by users.

File: lua/dap-python.lua (hunk 1 above)
Comment (theHamsta) [reply lua/dap-python.lua]: But then it should be able to run classes.

Comment (mfussenegger) [reply lua/dap-python.lua]: Ah right, sorry. Can you revert it for now? I'll then merge this - can be followed up later.

**Comment by mfussenegger** [comment]
> But the query files could still be in the standard directory to be overwritable by users and easier to edit.

Sounds good to me. Do you want to create a PR for that?

**Comment by theHamsta** [comment]
> Sounds good to me. Do you want to create a PR for that?

After this PR.

File: lua/dap-python.lua (hunk 1 above)
Comment (theHamsta) [reply lua/dap-python.lua]: @mfussenegger don't you think I should revert that change until nvim-dap-python properly supports the detection of test classes versus normal ones?

**Comment by theHamsta** [comment]
> I went back and forth trying to decide and in the end decided to make it language specific.

Even if it should be language specific the code would profit from separating the discovery of test methods via tree-sitter from the test launch.

File: lua/dap-python.lua (hunk 1 above)
Comment (mfussenegger) [reply lua/dap-python.lua]: > don't you think I should revert that change until nvim-dap-python properly supports the detection of test classes versus normal ones?

Yes that was what I meant.

Comment (theHamsta) [reply lua/dap-python.lua:5]: done

File: lua/dap-python.lua [hunk 2]
```diff
@@ -133,21 +136,45 @@ function M.test_method(opts)
...
+      })
+    else
+      print('No suitable test method found')
+    end
+  elseif test_runner == 'pytest' then
+    if closest_function then
+      local path = vim.fn.expand('%:p')
+      -- TODO: execution with class
+      -- local fqn = table.concat({path, classname, closest_function}, '::')
+      local fqn = table.concat({path, closest_function}, '::')
```
Comment (theHamsta) [inline lua/dap-python.lua:162]: Debugging a whole file would just be '{path}'

**Pull request merged**